sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,777 @@
+            else:
+                # This is only needed for compatibility for repeated scans.
+                labels[entry] = -1   # Noise points
+                is_core[entry] = 0
+        else:
+            labels[entry] = cluster_id
+            if core_distances[entry] <= eps:
+                is_core[entry] = 1   # True
+            else:
+                is_core[entry] = 0   # False
+    return np.arange(n_samples)[is_core], labels
+
+
+def _extract_optics(ordering, reachability, maxima_ratio=.75,
+                    rejection_ratio=.7, similarity_threshold=0.4,
+                    significant_min=.003, min_cluster_size_ratio=.005,
+                    min_maxima_ratio=0.001):
+    """Performs automatic cluster extraction for variable density data.
+
+    Parameters
+    ----------
+    ordering : array, shape (n_samples,)
+        OPTICS ordered point indices (`ordering_`)
+
+    reachability : array, shape (n_samples,)
+        Reachability distances calculated by OPTICS (`reachability_`)
+
+    maxima_ratio : float, optional
+        The maximum ratio we allow of average height of clusters on the
+        right and left to the local maxima in question. The higher the
+        ratio, the more generous the algorithm is to preserving local
+        minima, and the more cuts the resulting tree will have.
+
+    rejection_ratio : float, optional
+        Adjusts the fitness of the clustering. When the maxima_ratio is
+        exceeded, determine which of the clusters to the left and right to
+        reject based on rejection_ratio. Higher values will result in points
+        being more readily classified as noise; conversely, lower values will
+        result in more points being clustered.
+
+    similarity_threshold : float, optional
+        Used to check if nodes can be moved up one level, that is, if the
+        new cluster created is too "similar" to its parent, given the
+        similarity threshold. Similarity can be determined by 1) the size
+        of the new cluster relative to the size of the parent node or
+        2) the average of the reachability values of the new cluster
+        relative to the average of the reachability values of the parent
+        node. A lower value for the similarity threshold means less levels
+        in the tree.
+
+    significant_min : float, optional
+        Sets a lower threshold on how small a significant maxima can be.
+
+    min_cluster_size_ratio : float, optional
+        Minimum percentage of dataset expected for cluster membership.
+
+    min_maxima_ratio : float, optional
+        Used to determine neighborhood size for minimum cluster membership.
+
+    Returns
+    -------
+    core_sample_indices_ : array, shape (n_core_samples,)
+        The indices of the core samples.
+
+    labels_ : array, shape (n_samples,)
+        The estimated labels.
+    """
+
+    # Extraction wrapper
+    reachability = reachability/np.max(reachability[1:])
```

Comment:
This is not PEP8, add whitespaces around divide symbol.

On second thought, I wonder if this is should be normalized, since a non-normalized value would be consistent with the `eps` in DBSCAN, which is also not normalized. What do you think?

Reply:
fixed.

I think that your first intuition to normalize the parameter is correct. The other graph extraction parameters (`maxima_ratio`, `rejection_ratio`, `similarity_threshold`, `min_cluster_size_ratio`, and `min_maxima_ratio`) are all unitless, so I'd say that it makes sense to keep `significant_min` unitless as well. The `eps` parameter in DBSCAN/OPTICS has a pretty straightforward interpretation; if using `min_samples` of 5 and `eps` of 1-meter, then clusters are formed if 5 datapoints fall within a meter radius. For `significant_min`, it is a measure of difference in `reachability` distance between competing clusters...

Also, changing the parameter to be normalized appears to give better results-- github won't let me post the image for some reason, but if you compare the current plot output with what I posted yesterday, using the normalized `significant_min` gets a better cluster 'extent' when mapping the clusters :)